### ISSUE TITLE
qwen4exp: multi-seq split replay corrupts recurrent state when rs rollback is enabled (why the arch is excluded from llm_arch_supports_rs_rollback?)

### ISSUE THREAD

**Issue by netmana808** [ggml-org/llama.cpp#28019]
### Summary

`LLM_ARCH_QWEN4EXP` is excluded from `llm_arch_supports_rs_rollback()`, so `n_rs_seq` clamps to 0 for Qwen3.8-Flash-Next and hybrid-recurrent batched rollback never engages. Force-enabling it shows why: **`test-recurrent-state-rollback` fails `test_multi_seq_split_replay` with a logits max diff of 9.25** — single-seq checkpoint restore passes, multi-seq split replay corrupts state. If the exclusion is deliberate, this issue documents the concrete gap (and asks what state is missing snapshot coverage); if it's an oversight, the repro below should make it easy to chase.

### Repro (~2 min on top of loading the model)

On `master` @ `cc231cb0d`:

1. Add the arch to the rollback allowlist:

```diff
--- a/src/llama-arch.cpp
+++ b/src/llama-arch.cpp
@@ bool llm_arch_supports_rs_rollback(const llm_arch & arch) {
         case LLM_ARCH_QWEN35:
         case LLM_ARCH_QWEN35MOE:
+        case LLM_ARCH_QWEN4EXP:
         case LLM_ARCH_DEEPSEEK4:
```

2. Run the existing test against a Qwen3.8-Flash-Next GGUF (tested with an IQ4_XS quant, CPU backend):

```
CUDA_VISIBLE_DEVICES="" ./bin/test-recurrent-state-rollback \
  --model Qwen3.8-Flash-Next-IQ4_XS-00001-of-00003.gguf -ngl 0
```

3. Output:

```
main : recurrent rollback checkpoint restored successfully
test_multi_seq_split_replay : multi-seq split replay logits mismatch (max diff 9.25491, first at seq 0 pos 16)
```

A diff of 9.25 in the logits is state corruption, not numerical noise.

### Suspected cause

qwen4exp carries extra per-row recurrent state beyond the gated-delta-net state that the snapshot planes cover for the allowlisted archs:

- the **PLE conv history** (`ple_layer_ids: [2]` — the PLE module sits on a layer that is also a delta-net layer, so both keep conv state in the same recurrent row), and
- possibly **QSA indexer** state.

If those aren't widened into the `(1 + n_rs_seq)` snapshot groups, an equal-split replay would restore the GDN state but replay over stale conv/indexer state — consistent with the mismatch appearing early (pos 16) on seq 0.

### Why it matters

On hybrid-recurrent archs, server spec decode with `--parallel N` currently pays a large serialization cost without rollback snapshots. Measured on a GB10 (Grace-Blackwell, 128 GB unified, CUDA sm_121) with Qwen3.8-Flash-Next IQ4_XS + `--spec-type ngram-map-k4v --parallel 3`: three concurrent streams reach only ~50 tok/s aggregate on rewrite-style workloads vs ~66–70 with speculation disabled — speculation *inverts* under concurrency. Batched rollback is the structural fix, and it's blocked on this arch by the missing snapshot coverage.

Secondary observation while in this code: `common_params_speculative::need_n_rs_seq()` only arms `n_rs_seq` for the draft-model spec types (MTP/EAGLE3/DFLASH/DSPARK). Is there a reason the `ngram-*` types are excluded? They hit the same partial-acceptance rollback path.

### Offer

Happy to test patches on GB10 hardware — the repro takes ~2 minutes here and I can also run before/after server throughput numbers (`--parallel 3`, ngram spec) on the same box.

**Comment by Taudebug** [comment]
Data point from a newer master, where the exclusion this issue asks about is already gone — and an **earlier** check now fails than the one reported here.

On master @ `95ef7fc`, `LLM_ARCH_QWEN4EXP` is already in `llm_arch_supports_rs_rollback()` (added by #28123, `0eadefe`), so no patch is needed and `n_rs_seq` is not clamped. Running the unmodified test against a real Qwen3.8-Flash-Next checkpoint fails before it reaches `test_multi_seq_split_replay`:

```
$ ./bin/test-recurrent-state-rollback \
    --model Qwen3.8-Flash-Next-UD-IQ4_XS-00001-of-00003.gguf \
    -ngl 0 -c 2048 -fa on
main : dirty-ctx logits mismatch at position 6, token 0 (8.1648 != 5.47135)
$ echo $?
1
```

Same failure with the CUDA backend, same subtest, same position — only the values differ, by about what backend numerics would explain:

```
-ngl 999 -ncmoe 44 -ot per_layer_token_embd=CPU -c 2048 -fa on
main : dirty-ctx logits mismatch at position 6, token 0 (8.23151 != 5.62428)
```

The distinction from the report above: your run prints `main : recurrent rollback checkpoint restored successfully` and then fails in `test_multi_seq_split_replay`. Mine never prints that line and `multi_seq` never appears in the output at all — it exits at the dirty-context check, which sits earlier in `main()`. So on this build the single-seq checkpoint restore that passed for you is also failing.

The failing case is the one the test comments describe as loading a checkpoint into a context that "already has its own rollback state: groups 1..n_rs_seq hold a different prompt's history, and rs_idx[0] is non-zero at load time. The restore must wipe that state and still match." A diff of 8.23 vs 5.62 looks like the same class of state corruption you describe, not numerics.

Two things I have **not** established, in case they matter before anyone chases this:

- Whether this is a regression between `cc231cb0d` and `95ef7fc`, or whether the dirty-context check simply postdates your run. I cloned shallow (`--depth 200`), so `git log -S` attributes the test file to the boundary commit and I cannot date the check reliably.
- Whether `llama-server` reaches this path in ordinary single-slot use. With one continuous conversation in one slot the context may never hold a different prompt's history at restore time; reusing a slot across unrelated conversations would.

Environment: master `95ef7fc`, built with CUDA 13.3 + GCC 13.4, `-DCMAKE_CUDA_ARCHITECTURES=120` (RTX 5090, sm_120a), Ubuntu, glibc 2.43. The CPU-only run above uses none of the CUDA path and fails identically, so the toolchain does not look implicated. Model is the Unsloth `UD-IQ4_XS` split GGUF.

Happy to bisect or test a patch if that would help.

**Comment by netmana808** [comment]
Thanks — and first, the original ask is resolved: `LLM_ARCH_QWEN4EXP` is in `llm_arch_supports_rs_rollback()` on master, added by 0eadefeb (#28123, 2026-09-01). Worth noting that commit landed the *mechanism* I guessed at, not just the enum entry, and says so directly:

> The recurrent cache already holds n_rs_seq + 1 snapshot planes and the delta net writes its SSM state into them, but `build_conv_state_at` wrote a single plane, so a rollback restored a convolution history that was never captured. It now writes one snapshot per slot, each ending one token earlier, **for the delta net QKV convolution and for the PLE convolution alike.**

### Your question 1: the dirty-context check does not postdate my run

I have a full clone, so I can settle this. The check is old and was present in my build:

```
$ git log --oneline -S "dirty-ctx" -- tests/test-recurrent-state-rollback.cpp | tail -1
255582687 llama + spec: MTP Support  (#22673)

$ git show cc231cb0d:tests/test-recurrent-state-rollback.cpp | grep -n "dirty-ctx logits mismatch"
384:                fprintf(stderr, "%s : dirty-ctx logits mismatch at position %d, token %d (%g != %g)\n",
```

Ordering is also identical across both revisions, confirming your read that it sits earlier in `main()`:

| | cc231cb0d (mine) | master (yours) |
|---|---|---|
| dirty-ctx mismatch | line 384 | line 413 |
| "checkpoint restored successfully" | line 391 | line 420 |
| `test_multi_seq_split_replay(...)` | line 396 | line 425 |

So the same assertion, in the same position, **passed** on cc231cb0d with the arch force-enabled and **fails** on 95ef7fc where it is natively enabled. "New check" is ruled out.

### Why it may have passed for me — and why that points at #28123 rather than a regression elsewhere

Per the quote above, before #28123 `build_conv_state_at` wrote a single plane, so the convolution history was *never captured* for this arch. When I force-enabled the arch on cc231cb0d I therefore had an arch declared rollback-capable whose conv state was not actually being snapshotted. The dirty-context subtest compares a replay against a restore; if neither path captured conv history, both are wrong in the same way and the comparison can agree. It then failed later, at `test_multi_seq_split_replay`, where the per-sequence planes do get exercised.

After #28123 there are real per-slot conv snapshots, so the dirty-context restore finally has state it can get wrong — and the case the test comments describe is exactly the one where `rs_idx[0]` is non-zero at load and groups `1..n_rs_seq` hold a different prompt's history, which the restore must wipe.

If that reading is right, this is an **incompleteness in #28123's restore path for the newly-real conv planes**, not an unrelated regression between the two commits — which would also explain why a bisect might land on #28123 while the check itself is much older. A diff of 8.23 vs 5.62 at position 6 is far too large for backend numerics, and your CPU-only run failing identically agrees.

### Your question 2: does `llama-server` reach this in ordinary use?

It depends entirely on the speculator, because that is what arms `n_rs_seq`. On master:

```cpp
uint32_t need_n_rs_seq() const {
    bool needs_rs_seq = std::any_of(types.begin(), types.end(), [&](auto t) {
        return t == COMMON_SPECULATIVE_TYPE_DRAFT_MTP || t == COMMON_SPECULATIVE_TYPE_DRAFT_EAGLE3
            || t == COMMON_SPECULATIVE_TYPE_DRAFT_DFLASH || t == COMMON_SPECULATIVE_TYPE_DRAFT_DSPARK;
    });
    return needs_rs_seq ? draft.n_max : 0u;
}
```

- **No speculation, or any `ngram-*` type:** `n_rs_seq == 0`, so the rollback planes are never allocated and this path is unreachable no matter how slots are reused. I have run Qwen3.8-Flash-Next in production with `--spec-type ngram-map-k4v` for a couple of weeks across two machines without tripping anything that looks like this.
- **`draft-mtp` (and eagle3/dflash/dspark):** `n_rs_seq == draft.n_max > 0`, so the planes exist, and a slot reused across unrelated conversations should reach precisely the dirty-context case.

Since #28123 exists specifically to make MTP viable on this arch (its numbers: 183 tok/s on code with the standalone MTP draft versus 108 without), the exposed population is MTP users with more than one conversation per slot — i.e. the normal multi-turn server case, not an exotic one. That seems worth weighing when deciding urgency.

Happy to test a patch. I have two GB10s (aarch64 Grace + Blackwell, sm_121, CUDA 13) and can run both the CPU-only and CUDA paths on a real Flash-Next checkpoint, which gives a different toolchain and backend from your sm_120a/x86 setup.
